lib/src/hidden_ranges/hidden_line_ranges.dart

```diff
@@ -7,18 +7,81 @@ class HiddenLineRanges with EquatableMixin {
   final int fullLineCount;
   final int visibleLineCount;
 
-  const HiddenLineRanges({
+  final List<int> _recoveredLines;
+  final Map<int, int> _cutLines;
+
+  factory HiddenLineRanges({
+    required List<LineNumberingBreakpoint> breakpoints,
+    required int fullLineCount,
+    required int visibleLineCount,
+  }) {
+    final recoveredLines = <int>[];
+    final cutLines = <int, int>{};
+    for (var i = 0; i <= fullLineCount; i++) {
+      final visibleLine = _mapLineToVisible(i, breakpoints);
+      if (visibleLine != null) {
+        recoveredLines.add(i);
+        cutLines[i] = visibleLine;
+      }
+    }
+    return HiddenLineRanges._(
+      fullLineCount: fullLineCount,
+      visibleLineCount: visibleLineCount,
+      breakpoints: breakpoints,
+      recoveredLines: recoveredLines,
+      cutLines: cutLines,
+    );
+  }
+
+  const HiddenLineRanges._({
     required this.breakpoints,
     required this.fullLineCount,
     required this.visibleLineCount,
-  });
+    required List<int> recoveredLines,
+    required Map<int, int> cutLines,
+  })  : _recoveredLines = recoveredLines,
+        _cutLines = cutLines;
 
-  static const empty = HiddenLineRanges(
+  static const empty = HiddenLineRanges._(
     breakpoints: [],
     fullLineCount: 1,
     visibleLineCount: 1,
+    recoveredLines: [0],
+    cutLines: {},
   );
 
+  int? cutLineIndexIfVisible(int lineIndex) {
+    if (lineIndex < 0) {
```

Comment:
I think it's OK to drop these checks and throw if out of bounds. We now have tests for out of range lines, but we did support that only because with breakpoints it was easier to allow those. We can drop that support if it is now easier to only support valid ranges.